File: C_plus_plus/Gray_Code.cpp
class Solution {
public:
    vector<int> grayCode(int n) {
        if(!n) return vector<int>(1,0); 
        vector<bool> hash(1<<n,false);
        vector<int> result;
        result.push_back(0);
        hash[0] = true;
        int val = 0,tt;
        for(int i = 0 ; i < n ; ++i){
            if((val&(1<<i))>0) tt = val&(~(1<<i));
            else tt = val|(1<<i);
            if(!hash[tt]){
                val = tt;
                result.push_back(val);
                hash[val] = true;
                i = -1;
                continue;
            }
        }
        return result;
    }
};

File: java/Gray_Code.java
// Generate the number bit by bit
// When we arrive at the new bit , we shoud reverse the pre list and make or operation of 1 << i to each number to generate the 
// new list.
// Time and space are both O(2^n).

public class Solution {
    public List<Integer> grayCode(int n) {
        List<Integer> list = new ArrayList<Integer>();
        list.add(0);
        if (n == 0) return list;
        list.add(1);
        for (int i = 1; i < n; ++i) {
            int size = list.size();
            for (int j = size-1; j >= 0; --j)
                list.add(list.get(j) | 1 << i);
        }
        return list;
    }
}
